GS-Scaffold implementation

MrNeRF opened the pull request:
This is a draft implementation of scaffold-gs. It is still very raw, but I wanted to get the base running as quickly as possible.

The simple_trainer.py is already functional, and the rendering can be previewed with Viser. However, the densification algorithm (Algorithm 1) from [this CVPR 2024 supplemental document](https://openaccess.thecvf.com/content/CVPR2024/supplemental/Lu_Scaffold-GS_Structured_3D_CVPR_2024_supplemental.pdf) has not yet been implemented.

I spoke with Ruilong about working on this implementation. I'm sharing this draft to indicate that progress is being made (and someone is working on this implementation). I hope to finish a first complete GS-Scaffold implementation by tomorrow (but might be Friday, hahha :)

Here’s a small preview for now.

![Screenshot from 2024-09-18 14-13-02](https://github.com/user-attachments/assets/bb5fe5f4-e3ec-4bf5-a7d7-df3c0b25b0be)

MrNeRF commented:
@liruilong940607 I think it works quite well as of now but I haven't run the benchmarks yet.

What should we do about the file format? In theory, the mlp parameters need to be saved as well but that would create an own file format. Any thoughts?

MrNeRF commented:
I removed the pruning and use relocation. It seems to give better results.

https://github.com/user-attachments/assets/71b82a0b-9e88-4309-9fff-91c23b571408

liruilong940607 commented on examples/simple_trainer_scaffold.py:
```diff
@@ -0,0 +1,1117 @@
+    offsets = torch.zeros((N, strategy.n_feat_offsets, 3))
+
+    opacities = torch.logit(torch.full((N, 1), init_opacity))  # [N,]
+    params = [
+        # name, value, lr
+        ("anchors", torch.nn.Parameter(points), 0.0001),
+        ("scales", torch.nn.Parameter(scales), 0.007),
+        ("quats", torch.nn.Parameter(quats), 0.002),
+        ("opacities_mlp", strategy.opacities_mlp.parameters(), 0.002),
+        ("features", torch.nn.Parameter(features), 0.0075),
+        ("opacities", torch.nn.Parameter(opacities), 5e-2),
+        ("offsets", torch.nn.Parameter(offsets), 0.01 * scene_scale),
+        ("colors_mlp", strategy.colors_mlp.parameters(), 0.008),
+        ("scale_rot_mlp", strategy.scale_rot_mlp.parameters(), 0.004),
+    ]
+
+    splats = torch.nn.ParameterDict({n: v for n, v, _ in params}).to(device)
```
Having the MLPs to live inside the densification strategy class is a bit weird to me, as they (MLPs) are part of the representation of the scene, not the densification procedure. My proposal is to have a `ParameterDict` to host the anchor parameters and another `ModuleDict` to host the MLPs. Together they are the splats:
```
anchors: ParameterDict = ...
decoders: ModuleDict = ...
splats: {"anchors": anchors", "decoders": decoders}
```

Then, you could maintain two optimizers, one for the anchors and one for the MLPs. This way makes it easier for the densification strategy to update the parameters (you don't need to hard-code the key words in the `strategy/ops`):
```
optimizers = {"anchors": ..., "decoders": ...,}

strategy: ScaffoldStrategy = ...
strategy.step_pre_backward(params=splats["anchors"], optimizers=optimizers["anchors"])
...
```

For saving, I think you could just save splats as {"anchors": ..., "decoders": ....}.

MrNeRF replied:
I addressed your comments (hopefully). Please check.

liruilong940607 commented:
Great job @MrNeRF! I think saving in it's own file format is fine -- it is what it is

MrNeRF commented:
Thank you for your suggestions @liruilong940607 
I still play a little bit with the parameters and settings. If the measure are not somewhat flawed (and I need to confirm on more data), this could be the new sota.

I get now on the garden scene:
PSNR: 27.801, SSIM: 0.8678, LPIPS: 0.076 Time: 0.010s/image Number of GS: 1021525

MrNeRF commented:
Update: The last commit gives me
PSNR: 27.941, SSIM: 0.8705, LPIPS: 0.074 Time: 0.012s/image Number of GS: 1213319

liruilong940607 commented:
Hey this is insane. The default MCMC gives
PSNR: 26.983, SSIM: 0.8479, LPIPS: 0.113 Time: 0.059s/image Number of GS: 1000000

ichsan2895 commented:
I will help for benchmarking. Training now. Hopefully it runs without hassle

![image](https://github.com/user-attachments/assets/9aa40af8-8c90-48e5-a7d9-bbca23808228)

MrNeRF commented:
Here are my results. I think I overfitted a bit on the garden scene. What is the base line? According to nerfbaselines it is sota (psnr) on the following scenes: garden, counter, bonsai. The rest is pretty close. Only on the bike scene it is average.

In terms of lpips it is everywhere sota (is this comparable at all). I think the visual appearance is in general also better.

It is the fastest training and also mostly leading in memory consumption. 

**Eval Stats**
Scene | PSNR | SSIM | LPIPS  | Memory (GB) | Num ### GS (anchors)
-- | -- | -- | -- | -- | --
garden | 27.914 | 0.8718 | 0.0734 |  10.1964 | 1_072_915
bicycle | 25.174 | 0.7627 | 0.1599 | 15.8627 | 1_694_584
stump | 26.873 | 0.7793 | 0.1445  |14.5899 |  1_565_929
bonsai | 32.833 | 0.9462 | 0.1278  | 6.3332 | 669_712
counter | 29.749 | 0.9141 | 0.1518 | 3.7661 | 388_440
kitchen | 31.626 | 0.9312 | 0.0925 | 3.7940  | 387_015
room | 31.863 | 0.9219 | 0.1645  | 3.7305  | 387_108

ichsan2895 commented:
@MrNeRF 
Its strange, stump is not error with my experiment and using this commit [e00d134](https://github.com/nerfstudio-project/gsplat/pull/409/commits/e00d134d861cd3ca5916540b2b4176c44345bae9)

=== Eval Stats of stump ===
results/benchmark_scaffold/stump/stats/val_step29999.json
{"psnr": 26.83869171142578, "ssim": 0.7819640040397644, "lpips": 0.1407296359539032, "ellipse_time": 0.014932617545127869, "num_GS": 1560666}

MrNeRF commented:
Here is the latest state of this draft pr

| Scene   |    PSNR |   SSIM |   LPIPS |   Memory (GB) |   Num GS (anchors) |
|:--------|--------:|-------:|--------:|--------------:|-------------------:|
| garden  | 27.9297 | 0.8715 |  0.0728 |       10.8048 |            1141663 |
| bicycle | 25.2822 | 0.7641 |  0.163  |       16.5653 |            1766822 |
| stump   | 26.7358 | 0.7793 |  0.1428 |       14.6143 |            1550649 |
| bonsai  | 33.0422 | 0.9461 |  0.1273 |        6.6482 |             701940 |
| counter | 29.9382 | 0.9176 |  0.1464 |        3.9759 |             412276 |
| kitchen | 31.8974 | 0.9315 |  0.0913 |        3.6907 |             374534 |
| room    | 31.7386 | 0.9241 |  0.1553 |        3.3813 |             347669 |

ichsan2895 commented:
My result with `scaffold-gs` in commit [e00d134](https://github.com/nerfstudio-project/gsplat/pull/409/commits/e00d134d861cd3ca5916540b2b4176c44345bae9)
|   |  SCENE   |  PSNR  |  SSIM  | LPIPS |
|---|----------|-------:|-------:|------:|
| 0 | garden   | 27.922 | 0.873  | 0.074 |
| 1 | bicycle  | 25.250 | 0.769  | 0.157 |
| 2 | stump    | 26.839 | 0.782  | 0.141 |
| 3 | bonsai   | 33.036 | 0.948  | 0.125 |
| 4 | counter  | 29.947 | 0.919  | 0.144 |
| 5 | kitchen  | 32.004 | 0.932  | 0.092 |
| 6 | room     | 32.017 | 0.923  | 0.166 |

MrNeRF commented:
@ichsan2895 thx for testing. I think that confirms what is to be expected. Awesome!

ichsan2895 commented:
> @ichsan2895 thx for testing. I think that confirms what is to be expected. Awesome!

Don't worry @MrNeRF 

Now, I testing (again) with latest commit [139b521](https://github.com/nerfstudio-project/gsplat/pull/409/commits/139b52111c835db0785bcfd812417a96cb749652).

ichsan2895 commented:
Can it save to ply and view it using supersplat or mkkellogg gaussian viewer?

ichsan2895 commented:
My result with scaffold-gs in commit [139b521](https://github.com/nerfstudio-project/gsplat/pull/409/commits/139b52111c835db0785bcfd812417a96cb749652)

| SCENE    | PSNR   | SSIM   | LPIPS | VRAM (GB) | NUM_GS  |
|----------|--------|--------|-------|------------|---------|
| garden   | 27.917 | 0.873  | 0.075 | 12.583     | 1333404 |
| bicycle  | 25.165 | 0.766  | 0.162 | 18.581     | 1988100 |
| stump    | 26.805 | 0.780  | 0.145 | 15.343     | 1641045 |
| bonsai   | 33.043 | 0.947  | 0.126 | 7.146      | 756973  |
| counter  | 30.039 | 0.919  | 0.144 | 4.570      | 474202  |
| kitchen  | 32.175 | 0.934  | 0.091 | 3.985      | 404767  |
| room     | 31.924 | 0.926  | 0.153 | 3.916      | 407397  |
| AVERAGE  | 29.581 | 0.878  | 0.128 | 9.446      | 1000841 |

BTW, Can it save to ply and view it using supersplat or mkkellogg gaussian viewer?

MrNeRF commented:
> BTW, Can it save to ply and view it using supersplat or mkkellogg gaussian viewer?
@ichsan2895 
There is a way to do that. 
I thought gsplat is creating ply files but (if I don't missed it), its not doing it. So I voted for saving checkpoints to be consistent.

One can also remove the view dependency from the mlps and it also works well but the measures are a bit worse. I think this is something that should be supported by Nerfstudio? How do you usually convert the checkpoints to ply files?

ichsan2895 commented:
For default densification strategies & MCMC strategies, this snippet of code in this thread works well:
[Get .ply file after training?](https://github.com/nerfstudio-project/gsplat/issues/234)

MrNeRF commented:
I know how to do it. Is it like a big feature request? I can code it up in open3D.

But I will be gone for a week and not coding. I can do it thereafter for both, scaffold and the rest.

ichsan2895 commented:
> I know how to do it. Is it like a big feature request? I can code it up in open3D. I just want to get this here done for a review.

It is not big feature request, but it is HUGE feature request :1st_place_medal:

MrNeRF commented:
I think that's good for review now.

earth-bass commented:
Amazing work @MrNeRF! 
It doesn't seem like the current scaffold training is compatible with packed/sparse mode. Would be great if it had it integrated.

ichsan2895 commented:
Is this draft support multiGPU? I test it with this commit [e8d1207](https://github.com/nerfstudio-project/gsplat/pull/409/commits/e8d1207bcef4bbca9e8ea742fea3802bccf2a410)

I run this command:
```
SCENE_DIR="data/360_v2"
RESULT_DIR="results/benchmark_scaffold_2GPUs"
SCENE_LIST="garden bicycle stump bonsai counter kitchen room" # treehill flowers
RENDER_TRAJ_PATH="ellipse"

for SCENE in $SCENE_LIST;
do
    if [ "$SCENE" = "bonsai" ] || [ "$SCENE" = "counter" ] || [ "$SCENE" = "kitchen" ] || [ "$SCENE" = "room" ]; then
        DATA_FACTOR=2
    else
        DATA_FACTOR=4
    fi

    echo "Running $SCENE"

    # train without eval
    CUDA_VISIBLE_DEVICES=0,1 python3 examples/simple_trainer_scaffold.py --eval_steps 30000 --disable_viewer --data_factor $DATA_FACTOR \
        --render_traj_path $RENDER_TRAJ_PATH --steps_scaler 0.5 --packed \
        --data_dir data/360_v2/$SCENE/ \
        --result_dir $RESULT_DIR/$SCENE/
done
```
The error:
```
Traceback (most recent call last):
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/examples/simple_trainer_scaffold.py", line 1147, in <module>
    cli(main, cfg, verbose=True)
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/gsplat/distributed.py", line 344, in cli
    process_context.join()
  File "/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py", line 163, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 1 terminated with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/gsplat/distributed.py", line 295, in _distributed_worker
    fn(local_rank, world_rank, world_size, args)
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/examples/simple_trainer_scaffold.py", line 1103, in main
    runner = Runner(local_rank, world_rank, world_size, cfg)
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/examples/simple_trainer_scaffold.py", line 336, in __init__
    self.splats, self.optimizers = create_splats_with_optimizers(
  File "/workspace/GSPLAT_SCAFFOLD3/gsplat/examples/simple_trainer_scaffold.py", line 190, in create_splats_with_optimizers
    features = torch.zeros((N, cfg.feat_dim))
NameError: name 'cfg' is not defined
```

MrNeRF commented:
@ichsan2895 No. I can't test it properly as I don't have a multi gpu setup.

Either I remove it to not make it crash or I implement it blindly.

@earth-bass Thx. I'll look into it.

ichsan2895 commented:
My result with `scaffold-gs` in commit [e8d1207](https://github.com/nerfstudio-project/gsplat/pull/409/commits/e8d1207bcef4bbca9e8ea742fea3802bccf2a410)

| SCENE  | PSNR  | SSIM  | LPIPS | VRAM (GB) | NUM_GS   |
|--------|-------|-------|-------|-----------|----------|
| garden | 27.886| 0.872 | 0.075 | 12.533    | 1326524  |
| bicycle| 25.137| 0.765 | 0.164 | 18.463    | 1976070  |
| stump  | 26.810| 0.780 | 0.143 | 15.199    | 1625766  |
| bonsai | 32.948| 0.946 | 0.128 | 7.138     | 756253   |
| counter| 29.961| 0.919 | 0.145 | 4.692     | 487842   |
| kitchen| 32.103| 0.933 | 0.091 | 3.960     | 401787   |
| room   | 31.948| 0.925 | 0.155 | 3.941     | 410449   |
| AVERAGE| 29.542| 0.877 | 0.129 | 9.418     | 997813   |

This commit slightly worse than previous commit. But, it is still good and not significant.

ichsan2895 commented:
NEW SOTA! @MrNeRF :fire: 

Absgrad increase significantly the metrics at the cost increasing of Num_GS and VRAM.

Garden, bicycle, and stump are OOM even I use cloud GPU (RTX A6000 48 GB).

SCENE | PSNR | SSIM | LPIPS | VRAM (GB) | NUM_GS
-- | -- | -- | -- | -- | --
garden | 0.000 | 0.000 | 0.000 | 0.000 | 0
bicycle | 0.000 | 0.000 | 0.000 | 0.000 | 0
stump | 0.000 | 0.000 | 0.000 | 0.000 | 0
bonsai | 33.501 | 0.953 | 0.110 | 23.696 | 2539838
counter | 30.347 | 0.924 | 0.121 | 20.970 | 2229860
kitchen | 32.525 | 0.937 | 0.084 | 13.022 | 1355222
room | 32.318 | 0.931 | 0.136 | 19.273 | 2058410
AVERAGE | 18.384 | 0.535 | 0.064 | 10.994 | 1169047

MrNeRF commented:
Awesome! But it should not run oom. There is a vram intense part which is not optimally implemented. I can't fix it before next Sunday. 

But if someone wants to jump in, feel free.

MrNeRF commented:
It amazes me that abs grad improves the results by 0.4 - 0.5 dB @ichsan2895. That's absolutely freaking crazy.

It means also that this implementation easily destroys the leading methods on the 3DGS leaderboard in mipnerf360. Some scenes are outperformed by 1dB PSNR.

MrNeRF commented:
It amazes me that abs grad improves the results by 0.4 - 0.5 dB @ichsan2895. That's absolutely freaking crazy.

It means also that this implementation easily destroys the leading methods on the 3DGS leaderboard in mipnerf360. Some scenes are outperformed by 1dB PSNR.

ichsan2895 commented:
> It amazes me that abs grad improves the results by 0.4 - 0.5 dB @ichsan2895. That's absolutely freaking crazy.
> 
> It means also that this implementation easily destroys the leading methods on the 3DGS leaderboard in mipnerf360. Some scenes are outperformed by 1dB PSNR.

Yeah. Just fix the VRAM leak problem -> gets new SOTA -> create paper for next CVPR :laughing:

ichsan2895 commented:
> However, absgrad helps but I didn't evaluate it more systematically. Thanks for running the benchmark. Did you also inspect it visually. In my experience absgrad tends to introduce additional floaters.

So far, its good. It does not create extra unwanted floater in this kitchen scene. Needs to be inspect for another dataset.

Scaffold-GS
https://github.com/user-attachments/assets/62a6b76c-1161-4cfd-81f2-e8112241e896

Scaffold-GS with Absgrad
https://github.com/user-attachments/assets/288ff9bf-b155-4527-9326-a1a2a67cedd7

liruilong940607 commented:
Hey guys! It's super cool to see the the development here lead to new SOTA! However I would step back a little bit because the comparison here might be slightly unfair as this impl. produces more GSs, consumes more VRAM, and training time than the baselines. For almost all approaches of 3GDS, you could always get better performance by letting it converge to more GSs, with the price of spending more time and VRAM to train it. So I feel a super fair comparison should also consider training time & VRAM & number of GSs.

Btw, have you guys checked if this PR reimplements the performance of the original scaffold gs paper when following its original setup? I think when merging this pr, the default hyper parameters should be following the original paper, so that when user use it it should by default reproduce the paper's results. We then could support something more advanced like absgrad in the training script to get better performance, and log them in a README file or whatever.

MrNeRF commented:
There is currently a VRAM issue I need to resolve. It should be much less memory-hungry. But that won't impact the overrall structure of this pr.

The parameterization is different, as I spent some time fine-tuning it starting from what is given in the paper.

I can run some tests with the original parameters for comparison but thats already done on nerfbaselines. However, I noticed that the parameters and implementation from the paper do not align with key points highlighted in the paper. So what exactly should be compared? 

Also, note that I made some steps in the densification algorithm optional, as I found they performed worse. Most notably, I replaced the pruning with MCMC relocation. I found that pruning triggered a much stronger densification, which did not translate into higher quality. The relocation seems to dampen this behavior in favour of less anchors and better quality. I also removed dropout, which is not justified in the paper and does not help. There are a few other differences as well. For instance, the schedulers used in gsplat are already different which will give also a different performance. 

The voxelization, view-dependent MLPs, and anchor growth are implemented, which I believe are the main contributions of Scaffold-GS.

If I stay closer to the original Scaffold-GS (paper vs their code), the performance will be clearly worse. Absgrad can be off by default which I also did not activate in my tests. 

Comparing against mcmc can only be fair with the same final budget as what Scaffold-GS uses. But again, for every anchor 10 neural gaussians are spawned. Most scenes will run oom with this mcmc budget on a 4090RTX.

ichsan2895 commented:
Wow, one month later of this inactive thread.

If it not merged yet, please @MrNeRF don't remove this repo. It has big potential for giving the SOTA.

But I am curious how to export Gs-scaffold to PLY? Is it compatible with supersplat or mkkellogg gaussian-viewer?

arcman7 commented:
Sorry, just an interested bystander popping in-

Why can't this be merged? Even if the GPU requirements are heavy, using anything with GS-Scaffold is purely optional, so only people with the necessary hardware would use this right?

abrahamezzeddine commented:
@MrNeRF 

Hello. Thanks for your efforts. What is the current status for this? Is there any plan to finalise this? Thank you!

MrNeRF commented:
@abrahamezzeddine I don't have time to work on this currently. I also don't think the implementation is in line what the maintainers expect. You can take over if you like.

abrahamezzeddine commented:
> @abrahamezzeddine I don't have time to work on this currently. I also don't think the implementation is in line what the maintainers expect. You can take over if you like.

Thanks. May I ask what is remaining so I can understand the progress of this pull request?
